### PR TITLE
[Model][Spec Decode] Tap the pre-norm AttnRes mixture as the Kimi K3 DFlash aux state

### PR DESCRIPTION
**Pull request by rchalamala**
## Purpose

The DFlash drafter consumes auxiliary hidden states captured at a fixed set of
target layers. K3 captures the post-mixture stream, which is not what the
drafter was trained against: the AttnRes residual mixture is applied before
the layer norm, and the current capture site reads the value after it.

Tapping the pre-norm mixture instead recovers the stream the drafter expects.
This selects the correct weights to mix against depending on where the
tapped layer sits — the next layer's own AttnRes weights when one follows,
the model's output-side aggregation for the last layer on the final pipeline
stage, and a fall back to the running prefix for the last layer of a
non-final stage, where the consumer lives on the next rank and the
output-side weights do not exist there.

Gated behind `VLLM_KIMI_K3_AUX_ATTN_RES_STREAM` (default off) while it
settles, since it changes what the model feeds the speculator.

The one-shot log of which layers are tapped and which capture mode is active
is emitted from `_set_aux_hidden_state_layers`, the setup-time hook the
speculator calls once, rather than from inside `forward`. Setting an
`nn.Module` attribute inside a compiled forward pass risks a graph break or
recompile under `torch.compile`; doing it at setup time avoids that and is
also the natural place to log the tapped layer tuple.

I checked for duplicate/overlapping open PRs (`gh pr list --search` on "kimi k3 aux hidden state" and "attn_res eagle3"); no open PR touches the Kimi K3 auxiliary hidden state capture site. I used AI assistance (Cursor) to draft, test, and validate this change, and I reviewed every changed line before submitting.

## Test Plan

Unit:

```
pytest tests/models/kimi_k3/test_aux_attn_res_stream.py \
       tests/models/kimi_k3/test_eagle3.py \
       tests/models/kimi_k3/test_attn_res.py -v
```

End to end: Kimi K3 on 8×B300 at TP8, DFlash drafter at `num_speculative_tokens
16`, one concurrent user, single variable (the capture mode) on an otherwise
identical container. Acceptance is `1 + accepted/steps` from raw counter
deltas.

## Test Result

**Unit.** All pass:

```
tests/models/kimi_k3/test_aux_attn_res_stream.py .......          [7 passed]
tests/models/kimi_k3/test_eagle3.py ....                          [4 passed]
tests/models/kimi_k3/test_attn_res.py ......sss.....               [12 passed, 3 skipped]
```

The 3 skips are the multi-GPU (`nvidia-1/4/8`) parametrizations of
`test_attn_res`, not runnable on the single-GPU box used to verify this PR.
`test_aux_attn_res_stream.py` is new and covers the selection logic added
here: the consumer-layer branch, the last-layer/final-rank branch that uses
the model's output-side aggregation, the last-layer/non-final-rank fallback
to the running prefix, and that the feature is a no-op reproducing the
original `prefix_sum + hidden_states` when disabled.

One pre-existing test needed a fixture fix rather than a code change:
`test_eagle3.py::test_kimi_k3_uses_shared_eagle3_layer_configuration`'s
shared model stub predates this PR and didn't set `use_attn_res`, which the
new `_set_aux_hidden_state_layers` override now reads unconditionally. Fixed
the stub to set it, matching what the other two tests in that file already
do at their own call sites.

**End to end.** Mean accept length 4.7167 and 4.3948 with the capture on,
against 2.2913 and 2.6579 with it off. Paired across 32 matched rows: +1.903
accept length, treatment winning 31 of 32, p = 1.5e-8. An independent
implementation of the same idea reports +1.536 at a different operating
point (2.165 prefix-only against 4.454), which is the same effect measured
twice.

**How to read these numbers.** Some rows come from different containers
rather than uniformly from the same one, and the gain was measured with
prefix caching off; a follow-up with caching on read +0.083, inside the
within-arm spread. The flag is off by default for that reason — the
mechanism is right and the off-cache effect is large and well established,
but it should be re-measured on a cached lane before anyone depends on it
there.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by ZJY0516** [comment]
cc @mgoin @zixi-qi

**Comment by zixi-qi** [comment]
> cc @mgoin @zixi-qi

@ZJY0516 The change LGTM, this is probably more correct than our current approach

**Comment by rchalamala** [comment]
@zixi-qi Just rebased, let me know what you need from me.

**Comment by zixi-qi** [comment]
> @zixi-qi Just rebased, let me know what you need from me.

@ZJY0516 could you help with a stamp when you get a chance? Thanks!

**Review by ZJY0516 — APPROVED** [review]
Thanks for contribution

**Comment by yubofredwang** [comment]
The first two branches look right. I think the third one — the fallback to the
running prefix for the last layer of a non-final stage is wrong.

### Example

```
16 layers, attn_res_block_size = 4, pp = 2
  stage 0 = layers 0..7        stage 1 = layers 8..15
  aux capture id 8, i.e. the residual stream after layer 7
```

Capture id 8 is stage 0's `end_layer`, so this PR takes the fallback there.

What layer 8 actually reads, and what the same config captures at pp=1:

```python
sources = [bank[0], bank[1], prefix]          # cdiv(8, 4) = 2 blocks, plus the prefix
w       = softmax(score(s) for s in sources)  # score uses layer 8's res_norm / res_proj
feature = w[0]*bank[0] + w[1]*bank[1] + w[2]*prefix
```

What gets captured at pp=2:

```python
feature = prefix                              # i.e. w hard-coded to [0, 0, 1]
```

The mixture is a convex combination, so returning `prefix` alone is the same as
forcing the softmax one-hot onto the last source and discarding `bank[0]` and
`bank[1]`. Those aren't a remainder term — the weights are learned and generally
spread out. In a harness driving the real `attn_res` kernel (synthetic weights,
so illustrative), cosine similarity between the two expressions above ran 0.05
to 0.60.

The practical problem is that the same layer id now means two different things
depending on how the model is sharded, while the drafter at serving time always
consumes the first one. And it's silent: right shape, right dtype, no NaNs.

It needs `VLLM_KIMI_K3_AUX_ATTN_RES_STREAM=1` plus a capture id that coincides
with a stage end, so exposure grows with pp. For 93 layers with ids at
{8, 32, 48, 64, 88} under the default `get_pp_indices`, the first collision is at
pp=10 and from pp=40 on all five ids are stage ends.

### What stage 0 is actually missing

Only layer 8's two score vectors. Everything else is already local:

- `prefix` — local
- `bank` — local, and stage 0's `num_attn_res_blocks` is `cdiv(8, 4) = 2`, which
  is exactly layer 8's `prev_valid_blocks`, so it is already the right length
- `res_norm.weight` / `res_proj.weight` — on stage 1

Those two are one `hidden_size` vector each, unsharded and unquantised, about
28KB together at hidden 7168.

### Suggestion

Keep a copy of them on non-final stages. Redirect them in `load_weights` before
the PP filter drops them as belonging to a layer this rank does not own, then
use them for the boundary tap:

```python
else:
    # last layer of a non-final stage: the consumer is layer end_layer on the
    # next rank, but the prefix and bank it would read are both already here
    score_norm = self.boundary_attn_res_norm
    score_proj = self.boundary_attn_res_proj
    num_blocks = self.num_attn_res_blocks
```

Worth initialising them to NaN and checking at setup, so a checkpoint that never
supplies them fails loudly rather than exporting a mixture over uninitialised
memory.

I tried this and it reproduces the pp=1 capture bit for bit at every split point
(24 layers / block size 4 and 32 layers / block size 8, real kernel, hidden
7168). Happy to open a PR against your branch if that's useful.

One last thing: the test covering this branch currently asserts the fallback, so
it would pin the behaviour either way.

**Comment by njhill** [comment]
@rchalamala is this ready to merge now from your pov?

**Comment by rchalamala** [comment]
@njhill Yes

**Comment by rchalamala** [comment]
@njhill had to fix a small env var conflict, could you merge it now?
